tests/test_products.py

```diff
@@ -109,3 +117,101 @@ def test_skip_brew_tag_linking_for_buggy_products(requests_mock):
+            f"{settings.PRODSEC_DASHBOARD_URL}/product-definitions", text=(prod_defs.read())
+        )
+
+    update_products()
+
+    # Verify that a new ProductVariant called "2" was created
+    variant_2 = ProductVariant.objects.get(name="2")
+    assert variant_2
+    assert stream == variant_2.productstreams
+
+    # Verify that the original variant is still associated with the stream
+    assert stream == variant.productstreams
+
+    assert not mock_update_builds.called
+
+
+@pytest.mark.django_db
+@patch("corgi.tasks.prod_defs.slow_update_builds_for_variant.delay")
+def test_stream_variants_moved_to_new_stream(mock_update_builds, requests_mock):
+    # Creates ProductStream, "stream", with an existing ProductVariant called "1"
+    # The ps_update_stream loaded from proddefs-*.json below has a matching name "stream"
+    stream, variant = setup_product("stream")
+
+    # Verify that the original variant is associated with the stream
+    assert stream == variant.productstreams
+
+    # Sets up a CollectorErrataProductVersion with a matching brew tag
+    et_product = CollectorErrataProduct.objects.create(et_id=1, name="et_product_1")
+    et_version = CollectorErrataProductVersion.objects.create(
+        et_id=10, product=et_product, name="et_stream_1", brew_tags=["two_streams_with_same_tag"]
+    )
+    # This is the same variant as the original to the CollectorErrataProductVersion
+    CollectorErrataProductVariant.objects.create(et_id=100, name="1", product_version=et_version)
+
+    with open("tests/data/proddefs-update-variants-moved.json") as prod_defs:
+        requests_mock.get(
+            f"{settings.PRODSEC_DASHBOARD_URL}/product-definitions", text=(prod_defs.read())
+        )
+
+    update_products()
+
+    variant = ProductVariant.objects.get(name="1")
+    new_stream = ProductStream.objects.get(name="new_stream")
+    assert variant.productstreams == new_stream
+    # The variant is no longer associated with the original stream
+    assert not stream == variant.productstreams
+    assert mock_update_builds.called_with("1", ("new_stream", "stream"))
+
+
+@pytest.mark.django_db
+@patch("corgi.tasks.prod_defs.slow_update_builds_for_variant.delay")
+def test_product_variant_index_out_of_range(mock_update_builds, requests_mock):
+    setup_product()
+
+    et_product = CollectorErrataProduct.objects.create(et_id=1, name="et_product_1")
+    # Sets up 2 CollectorErrataProductVersions with matching brew tags
+    # The first one has no matching stream in product_definitions.json
+    et_version_2 = CollectorErrataProductVersion.objects.create(
+        et_id=12, product=et_product, name="et_stream_2", brew_tags=["two_streams_with_same_tag"]
+    )
+    et_version = CollectorErrataProductVersion.objects.create(
+        et_id=11, product=et_product, name="et_stream_1", brew_tags=["two_streams_with_same_tag"]
+    )
+    # This is a new variant for et_stream_2, but should get associated with the original stream
```

Comment:
I think this test was confusing because Variant "1" was first created by the setup_product() call, and the new streams in the proddefs data shared a product with that function as well. Therefore variant "1" is moved twice, from the stream created in setup_product to "stream", and then again to "new_stream". Variant "2" is also moved once. I added some extra comments and asserts to better demonstrate this.